NumericInput: Button Increment Rounding

RolandLocke opened the pull request:
**Pull Request Description**

Added:
- Form input now follows below rules to help users better understand the behavior:
```
If the component is in its initial state (i.e. has not been manually edited and no increment or decrement buttons have been clicked):
  If an increment button larger than the current value is clicked (e.g. value 60, click +100)
    The value is set to the next higher value of that increment (e.g. value -> 100).
  otherwise (e.g. value 60, click +10)
    The value is incremented based on the button that is clicked. (e.g. value -> 70)
otherwise
  The value is incremented or decremented based on the button clicked.
```
- Button disabling and styling when decrementing the input will drop it below the min value. 
- The input form now has `initialValueChanged` data, indicating when the user has first interacted with the form input.

Closes #204 

---

**Licensing Certification**

FarmData2 is a [Free Cultural Work](https://freedomdefined.org/Definition) and all accepted contributions are licensed as described in the LICENSE.md file. This requires that the contributor holds the rights to do so. By submitting this pull request **I certify
/ that I satisfy the terms of the [Developer Certificate of Origin](https://developercertificate.org/)** for its contents.

Commit:
Added button disabling and styling when decrementing the input will drop it below the min value. The input now knows when it has left default state. Incrementing the input by a value greater than the default value will completely replace the value.

Commit:
Changes defaultValueChanged variable to initialValueChanged. Added valueChanged method which updates the value via v-on:update:model-value call in the forminput template. Now behavior is as expected when the user types values in the form input.

Commit:
began adding tests to the Numeric Input Component

braughtg commented:
Had a look at this this morning and noticed one thing with regard to the minimum value.  When playing with this on the examples page, with the minValue to 0, the small decrement button is disabled (correct). But if I then click the small increment button making the value 1, the small decrement button is still not enabled (incorrect).  If I click the small increment button again, making the value 2, then the small decrement button is enabled.  So there seems to be an off by one error in there somewhere.  Similar behavior occurs when the minValue is not 0.  I did not test the medium or large buttons.

braughtg committed:
adds a numericValue data attribute to enable the component to use
the current value of the text field in its computations when running
in component tests.

braughtg commented:
As we discussed in person yesterday, you were seeing an issue where the small increment button was not becoming enabled as expected when the value was increased when running in a component test.  However, the behavior was correct when looking at the `NumericInput` component on the example page.  

I have figured out what the issue is and pushed some changes to this branch.  Briefly, the issue is that the changes you had made to the component rely on `this.value`.  When used in a page the value `prop` is `v-model`ed to a `data` element in the page.  This way, when the component emits the `update:value` event, the value goes to the page and then comes back to the component via the `v-model`.  The new value is then seen inside of the component and all works as expected. 

However, when running in a component test, there is no `v-model` that feeds the new value back into the component. So the component is only seeing the initial value of the `value` `prop` that is set in the test.  I.e. changing the value in the text field does emit the `update:value` event, but the page does not in turn update the associated `prop`.  So when the component looks at the `value` `prop` it is still seeing the initial value. 

The fix was to create an internal numeric value in `data` that can then be used to test things numerically against the current value of the field.   Check the diff for the commit that I pushed for the details.

Commit:
added final touches to the button disabling

Commit:
added tests for the input behavior

RolandLocke commented:
> The fix was to create an internal numeric value in `data` that can then be used to test things numerically against the current value of the field. Check the diff for the commit that I pushed for the details.

I moved the assignment of this variable to the `valueAsString()` method as it applies to text and form inputs in addition to inc/dec button presses.

RolandLocke commented:
@braughtg I noticed that the rounded corners on the buttons weren't being rendered on the rightside. 

Before:
![image](https://github.com/user-attachments/assets/4c24a28d-5a11-4f41-aef2-76cdfb822322)

After:
![image](https://github.com/user-attachments/assets/b564945d-4ceb-4516-b87b-27e789ba754d)

This was due to a `class="has-validation"` class assignment missing from the `inputGroup` element. Discussion [linked](https://github.com/bootstrap-vue/bootstrap-vue/issues/5314). I fixed it on this page, but it looks like its an issue on almost every input with a button attached:
![image](https://github.com/user-attachments/assets/60001897-0b8b-4c9d-bbf0-d9e5d26ebeea)

braughtg commented:
> This was due to a `class="has-validation"` class assignment missing from the `inputGroup` element. Discussion [linked](https://github.com/bootstrap-vue/bootstrap-vue/issues/5314). I fixed it on this page, but it looks like its an issue on almost every input with a button attached:

Good find and fix.  Would you be willing to open an issue and create a PR to fix this in the `SelectorBase` component as well?

braughtg commented on components/NumericInput/NumericInput.vue:
There seems to be a disagreement between naming and behavior in the functions used for disabling the increment buttons.  For example, the `numeric-increase-small` button is enabled/disabled using the `diableLargeInc` computed property which doesn't make semantic sense:
```
v-bind:variant="disableLargeInc ? 'outline-secondary' : 'outline-success'
```

It appears that this works because the `disableLargeInc` property uses `incDecValues[0]` which is actually the small increment/decrement value.  

A similar issue exists in the other functions used for the medium and large increment buttons.

braughtg commented on components/NumericInput/NumericInput.behavior.comp.cy.js:
```diff
@@ -61,6 +61,166 @@ describe('Test the NumericInput component behavior', () => {
+      });
+  });
+
+  it('Input of a larger value replaces the initial value', () => {
```
This test and the next one have the same `it` message.  They should be edited to be descriptive of what they are specifically testing, without getting too long.

braughtg replied:
Also seems like there should be a similar test for the medium increase button.

RolandLocke replied:
I've updated both these tests to have a single increment value either larger or smaller than the initial value. This tests all cases as the logic to determine if the initial value should be replaced all happens in the the same function.  It simply depends on the `incDecValues[]` defined by the form.

braughtg reviewed (changes requested):
Looking good.  A few comments describing some improvements that should be made.

In addition, I added a comment to the issue asking about the behavior of the decrement buttons on the first click.  Should they behave similarly to the way the increment buttons behave?

braughtg commented on components/NumericInput/NumericInput.behavior.comp.cy.js:
```diff
@@ -61,6 +61,166 @@ describe('Test the NumericInput component behavior', () => {
+      });
+  });
+
+  it('Input of a larger value replaces the initial value', () => {
```
Should there also be a test for the small button for when there are decimals in the value?  For example, min value is 0, current value is 0.5 and the small increment is 1.

RolandLocke replied:
I'm unsure if it is necessary, but I've added a test for a decimal value just for safety.

Commit:
removed has-validation class to move to appropriate issue

RolandLocke commented:
> Good find and fix. Would you be willing to open an issue and create a PR to fix this in the `SelectorBase` component as well?
 
See new [pull request ](https://github.com/FarmData2/FarmData2/pull/392)+ [issue](https://github.com/FarmData2/FarmData2/issues/391)

Commit:
fixed computed property naming schemes and added/changed appropriate tests

braughtg committed:
fixes dependent tests in SoilDisturbance component

braughtg commented:
I pushed a few changes to the `SoilDisturbance` tests because they relied on clicking buttons that are now disabled.

braughtg committed:
patches some tray seding tests for new inc/dec button behavior

braughtg commented:
Just pushed some changes to the `TraySeeding` tests to address similar issues.
